core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+/**
+ * A `ZRef[EA, EB, A, B]` is a mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
+ * a volatile `var`, augmented with atomic operations, which make it useful as a
+ * reasonably efficient (if low-level) concurrency primitive.
+ *
+ * {{{
+ * for {
+ *   ref <- Ref.make(2)
+ *   v   <- ref.update(_ + 3)
+ *   _   <- console.putStrLn("Value = " + v) // Value = 5
+ * } yield ()
+ * }}}
+ */
+trait ZRef[+E, A, B] extends Serializable { self =>
+
+  /**
+   * Reads the value from the `Ref`.
+   */
+  def get: IO[E, B]
+
+  /**
+   * Writes a new value to the `Ref`, with a guarantee of immediate
+   * consistency (at some cost to performance).
+   */
+  def set(a: A): IO[E, Unit]
+
+  /**
+   * Writes a new value to the `Ref` without providing a guarantee of
+   * immediate consistency.
+   */
+  def setAsync(a: A): IO[E, Unit]
+
+  /**
+   * Apply a partial function `pf` to the content of the Ref, if `pf` is not defined over the input it will fail on `EB` with `None`
+   */
+  // final def collect[C](pf: PartialFunction[B, C]): ZRef[EA, Option[EB], A, C] =
+  //   new ZRef[EA, Option[EB], A, C] {
+  //     def set(a: A): IO[EA, Unit] = self.set(a)
+  //     def get: IO[Option[EB], C] = self.get.foldM(
+  //       error => ZIO.fail(Some(error)),
+  //       value => pf.lift(value).fold[IO[Option[EB], C]](ZIO.fail(None))(ZIO.succeed(_))
+  //     )
+  //     def setAsync(a: A): zio.IO[EA, Unit] = self.setAsync(a)
+  //     private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, C] =
+  //       self.unsafeUpdate(f).foldM(error => ZIO.fail(error), value => ZIO.succeed(f(value).asInstanceOf[C]))
+  //     private[zio] def unsafeModify[E, D](f: Any => (Any, Any)): zio.IO[E, D] = self.unsafeModify(f)
+  //   }
+
+  // /**
+  //  * Replace the content of the Ref with `c`
+  //  */
+  // final def const[C](c: C): ZRef[EA, EB, A, C] = map(_ => c)
+
+  // /**
+  //  * Transforms elements that are set into this Ref with a function.
+  //  */
+  // final def contramap[C](ca: C => A): ZRef[EA, EB, C, B] =
```

Comment:
Now you should be able to write `contramap` 🎉

Reply:
yay even the test that failed because it was unsound passes now